src/cdi_CPEloss/Tabular_CP_Eloss.cc

```diff
@@ -0,0 +1,270 @@
+  using std::stod;
+  using std::stoi;
+
+  model_type = rtt_cdi::CPModelType::TABULAR_ETYPE;
+
+  file.open(filename);
+  std::ostringstream file_error;
+  file_error << "Error opening DEDX file \"" << filename << "\"";
+  Insist(file.is_open(), file_error.str());
+
+  std::vector<std::string> line_entries;
+  uint32_t nlines;
+  constexpr uint32_t max_entries =
+      6; // This is a statement about the file format, maximum of six entries per row.
+
+  line_entries = read_line(); // ZAID
+  int32_t projectile_zaid_file = stoi(line_entries[0]);
+  Require(projectile.get_zaid() == projectile_zaid_file);
+
+  read_line(); // Z, A, mass
+
+  line_entries = read_line(); // Number of bins for energy, density, temperature
+  n_energy = stoi(line_entries[0]);
+  n_density = stoi(line_entries[1]);
+  n_temperature = stoi(line_entries[2]);
+
+  line_entries =
+      read_line(); // Bin spacing for energy, density, temperature (log)
+  d_log_energy = 1. / stod(line_entries[0]);
+  d_log_density = 1. / stod(line_entries[1]);
+  d_log_temperature = 1. / stod(line_entries[2]);
+
+  // Get first energy support point
+  nlines =
+      static_cast<int>(std::ceil(static_cast<double>(n_energy) / max_entries));
+  line_entries = read_line();
+  min_log_energy = stod(line_entries[0]);
+  for (uint32_t n = 0; n < nlines - 1; n++) {
+    read_line();
+  }
+  energies.resize(n_energy);
+  for (uint32_t n = 0; n < n_energy; n++) {
+    energies[n] = exp(min_log_energy + n * d_log_energy);
+  }
+
+  // Get first density support point
+  nlines =
+      static_cast<int>(std::ceil(static_cast<double>(n_density) / max_entries));
+  line_entries = read_line();
+  min_log_density = stod(line_entries[0]);
+  for (uint32_t n = 0; n < nlines - 1; n++) {
+    read_line();
+  }
+  densities.resize(n_density);
+  for (uint32_t n = 0; n < n_density; n++) {
+    densities[n] = exp(min_log_density + n * d_log_density);
+  }
+
+  // Get first temperature support point
+  nlines = static_cast<int>(
+      std::ceil(static_cast<double>(n_temperature) / max_entries));
+  line_entries = read_line();
+  min_log_temperature = stod(line_entries[0]);
+  for (uint32_t n = 0; n < nlines - 1; n++) {
+    read_line();
+  }
+  temperatures.resize(n_temperature);
+  for (uint32_t n = 0; n < n_temperature; n++) {
+    temperatures[n] = exp(min_log_temperature + n * d_log_temperature);
+  }
+
+  std::vector<double> stopping_data_1d(n_energy * n_density * n_temperature);
+
+  bool target_found = false;
+  nlines = static_cast<int>(std::ceil(
+      ((double)n_energy * n_density * n_temperature) /
+      max_entries)); // The number of lines taken up by stopping power data for one target
+  if (target.get_zaid() == -1) {
+    // Target is free electrons
+    target_found = true;
+    uint32_t nentry = 0;
+    for (uint32_t n = 0; n < nlines; n++) {
+      line_entries = read_line();
+      for (std::string entry : line_entries) {
+        stopping_data_1d[nentry] = stod(entry);
```

Comment:
Please double check, but I think you can eliminate line 121 by saying `stopping_data_1d[nentry++] = stod(entry);`

Reply:
I didn't think this would work but it does (I checked in a separate test code, and confirmed online that `expr++` returns `expr` before incrementing)